libvast/vast/chunk.hpp

```diff
@@ -62,22 +65,32 @@ class chunk : public caf::ref_counted {
   /// @pre `size > 0 && static_cast<bool>(deleter)`
   static chunk_ptr make(size_type size, void* data, deleter_type deleter);
 
-  /// Construct a chunk from a container of bytes.
-  /// @param xs The container of bytes.
+  /// Construct a chunk from a std::vector of bytes.
+  /// @param xs The std::vector of bytes.
   /// @returns A chunk pointer or `nullptr` on failure.
   /// @pre `std::size(xs) != 0`
-  template <class Container,
-            class = std::enable_if_t<detail::is_container<Container>>>
-  static chunk_ptr make(Container xs) {
-    static_assert(sizeof(typename Container::value_type) == 1,
-                  "chunks only support byte containers");
+  // FIXME: Make a type whitelist. This optimization only makes sense for owning
+  // containers of bytes that implement move constructors. (i.e. we want to
+  // allow vector<char>, vector<byte>, flatbuffer, etc. but not string_view,
+  // vast::span, etc.)
+  template <typename Byte>
+  static chunk_ptr make(std::vector<Byte>&& xs) {
+    static_assert(sizeof(Byte) == 1);
     VAST_ASSERT(std::size(xs) != 0);
-    auto ys = std::make_shared<Container>(std::move(xs));
+    auto ys = std::make_shared<std::vector<Byte>>(std::move(xs));
     auto deleter = [=]() mutable { ys.reset(); };
-    auto data = std::data(*ys);
-    // The deleter won't touch the data.
-    using mutable_data = std::decay_t<decltype(*data)>*;
-    return make(std::size(*ys), const_cast<mutable_data>(data), deleter);
+    return make(ys->size(), ys->data(), deleter);
+  }
+
+  // Construct a chunk by copying a range of bytes.
+  template <typename Byte>
+  static chunk_ptr copy(size_type size, const Byte* data) {
```

Comment:
```suggestion
  static chunk_ptr copy(span<const byte> bytes) {
```

I think this interface gives more flexibility because there exist a lot of utility functions to convert contiguous things into a `span<const byte>`. For example, we also have `as_bytes(T)` for that.

Reply:
Good point, that also fixes the "wrong" order of size and pointer.